src/osd/OSD.cc

```diff
@@ -8276,7 +8276,20 @@ void OSD::handle_pg_create(OpRequestRef op)
 	       << dendl;
       continue;
     }
-
+    {
+      RWLock::RLocker l(pg_map_lock);
+      const auto max_pgs_per_osd = cct->_conf->get_val<uint64_t>("osd_pg_max_per_osd");
+      if (max_pgs_per_osd > 0 && pg_map.size() > max_pgs_per_osd) {
+	dout(4) << __func__ << ": skipping " << pgid << ": pg number "
+		<< pg_map.size() << " > "
+		<< max_pgs_per_osd
+		<< dendl;
+	want_to_create++;
+	continue;
+      } else {
+	want_to_create = 0;
+      }
+    }
```

Comment:
This is only the first half of the problem (and the easier of the two :).  The real challenge is PG creation due to a pgnotify from a peer.  We want to prevent existing PGs from rebalancing onto the node.

For that, I think we need to maintain a set<spg_t> want_to_create, and when the pg could drops, retrigger peering for that PG (probably by twiddling pg_temp to induce a new interval).  Your want_to_create should probably be renamed want_to_create_mon and the other one want_to_create_peer or something so they can be distinguished.  For the mon case a simple counter is enough, but for the peer case we need to know the identify of the pg.  (Hmm, we probably also need to track the epoch we decided to start dropping the peering messages, and keep dropping them until the new interval happens, or else we risk thoroughly confusing the peering code?)